File: Lecture4Exercise/chr20_maf_hwe_compare.R
library(data.table)
library(ggplot2)
library(dplyr)

# Load data
eur_freq <- fread("eur_stats.afreq") %>% mutate(Pop = "EUR")
afr_freq <- fread("afr_stats.afreq") %>% mutate(Pop = "AFR")
eur_hwe <- fread("eur_stats.hardy") %>% mutate(Pop = "EUR")
afr_hwe <- fread("afr_stats.hardy") %>% mutate(Pop = "AFR")

# Combine datasets
freq_all <- rbind(eur_freq, afr_freq)
hwe_all <- rbind(eur_hwe, afr_hwe) %>% 
  filter(P > 0) %>% 
  mutate(negLogP = -log10(P))

# Start PDF output
pdf("chr20_maf_hwe_compare.pdf", width = 10, height = 7)

# Plot 1: MAF Comparison (Density)
p1 <- ggplot(freq_all, aes(x = ALT_FREQS, fill = Pop)) +
  geom_density(alpha = 0.4) +
  theme_minimal() +
  labs(title = "MAF Density Comparison: EUR vs AFR (Chr 20)",
       x = "Alternate Allele Frequency", y = "Density") +
  scale_fill_manual(values = c("EUR" = "#377eb8", "AFR" = "#e41a1c"))

print(p1)

# Plot 2: HWE Comparison (-log10 P)
p2 <- ggplot(hwe_all, aes(x = negLogP, fill = Pop)) +
  geom_histogram(bins = 50, alpha = 0.7, position = "identity") +
  facet_wrap(~Pop) +
  geom_vline(xintercept = 6, linetype = "dashed") +
  theme_light() +
  labs(title = "Hardy-Weinberg Equilibrium Comparison",
       subtitle = "Dashed line at -log10(P) = 6",
       x = "-log10(P-value)", y = "SNP Count") +
  scale_fill_manual(values = c("EUR" = "#377eb8", "AFR" = "#e41a1c"))

print(p2)

dev.off()

File: Lecture4Exercise/chr20_maf_hwe_compare.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.backends.backend_pdf import PdfPages

# Load datasets
eur_f = pd.read_csv("eur_stats.afreq", sep='\t').assign(Pop='EUR')
afr_f = pd.read_csv("afr_stats.afreq", sep='\t').assign(Pop='AFR')
eur_h = pd.read_csv("eur_stats.hardy", sep='\t').assign(Pop='EUR')
afr_h = pd.read_csv("afr_stats.hardy", sep='\t').assign(Pop='AFR')

# Combine and calculate -log10(P)
freq_all = pd.concat([eur_f, afr_f])
hwe_all = pd.concat([eur_h, afr_h])
hwe_all = hwe_all[hwe_all['P'] > 0].copy()
hwe_all['neg_log10_p'] = -np.log10(hwe_all['P'])

with PdfPages('chr20_maf_hwe_compare.pdf') as pdf:
    # Page 1: MAF Comparison
    plt.figure(figsize=(10, 6))
    sns.kdeplot(data=freq_all, x="ALT_FREQS", hue="Pop", fill=True, palette="Set1", common_norm=False)
    plt.title("Allele Frequency Comparison (EUR vs AFR)")
    plt.xlabel("Alternate Allele Frequency")
    plt.ylabel("Density")
    pdf.savefig()
    plt.close()

    # Page 2: HWE Comparison
    fig, axes = plt.subplots(1, 2, figsize=(12, 5), sharey=True)
    for i, pop in enumerate(['EUR', 'AFR']):
        data = hwe_all[hwe_all['Pop'] == pop]
        sns.histplot(data['neg_log10_p'], bins=50, ax=axes[i], color='crimson' if pop=='AFR' else 'steelblue')
        axes[i].set_title(f"HWE Distribution: {pop}")
        axes[i].axvline(6, color='black', linestyle='--')
        axes[i].set_xlabel("$-log_{10}(P)$")
    
    plt.tight_layout()
    pdf.savefig()
    plt.close()

print("Python PDF Report Generated.")
